src/taxes/components/TaxCountryDialog/TaxCountryDialog.tsx

```diff
@@ -0,0 +1,119 @@
+import {
+  Dialog,
+  DialogActions,
+  DialogContent,
+  InputAdornment,
+  TextField
+} from "@material-ui/core";
+import VerticalSpacer from "@saleor/apps/components/VerticalSpacer";
+import { CountryFragment } from "@saleor/graphql";
+import { ChangeEvent } from "@saleor/hooks/useForm";
+import { useLocalSearch } from "@saleor/hooks/useLocalSearch";
+import { buttonMessages } from "@saleor/intl";
+import { Button, DialogHeader, SearchIcon } from "@saleor/macaw-ui";
+import { taxesMessages } from "@saleor/taxes/messages";
+import React from "react";
+import { FormattedMessage, useIntl } from "react-intl";
+
+import { useStyles } from "./styles";
+import { TaxCountryDialogLine } from "./TaxCountryDialogLine";
+
+interface TaxCountryDialogProps {
+  open: boolean;
+  countries: CountryFragment[];
+  onConfirm: (countries: CountryFragment[]) => void;
+  onClose: () => void;
+}
+
+export interface CountryFragmentWithState extends CountryFragment {
+  checked: boolean;
+}
+
+export const TaxCountryDialog: React.FC<TaxCountryDialogProps> = ({
+  open,
+  countries,
+  onConfirm,
+  onClose
+}) => {
+  const classes = useStyles();
+  const intl = useIntl();
+
+  const [countriesWithState, setCountriesWithState] = React.useState<
+    CountryFragmentWithState[]
+  >([]);
+  React.useEffect(
+    () =>
+      setCountriesWithState(
+        countries.map(country => ({ checked: false, ...country }))
+      ),
+    [countries]
+  );
+  const handleChange = React.useCallback(
+    (e: ChangeEvent) => {
+      const countriesUpdate = [...countriesWithState];
+      countriesUpdate.find(country => country.code === e.target.name).checked =
+        e.target.value;
+      setCountriesWithState(countriesUpdate);
```

Comment:
You dereferenced countriesWithState, but objects stay the same.

```suggestion
      setCountriesWithState(prevCountries => prevCountries.map(country => country.code === e.target.name ? {
        ...country,
        checked: e.target.value
      } : country))
```